src/sinks/aws_s3/integration_tests.rs

```diff
@@ -406,6 +406,51 @@ async fn s3_healthchecks_invalid_bucket() {
         .is_err());
 }
 
+#[tokio::test]
+async fn s3_flush_on_exhaustion() {
+    let cx = SinkContext::new_test();
+
+    let bucket = uuid::Uuid::new_v4().to_string();
+    create_bucket(&bucket, false).await;
+
+    let config = config(&bucket, 10); // batch size is ten
+    let prefix = config.key_prefix.clone();
+    let service = config.create_service(&cx.globals.proxy).await.unwrap();
+    let sink = config.build_processor(service).unwrap();
+
+    let (lines, _events) = random_lines_with_stream(100, 2, None); // only generate two events (less than batch size)
+
+    let events = lines.clone().into_iter().enumerate().map(|(i, line)| {
+        let mut e = LogEvent::from(line);
+        let i = if i < 10 {
+            1
+        } else if i < 20 {
+            2
+        } else {
+            3
+        };
+        e.insert("i", i.to_string());
+        Event::from(e)
+    });
+
+    // Here, we validate that the s3 sink flushes when its source stream is exhausted
+    // by giving it a number of inputs less than the batch size, verifying that the
+    // outputs for the in-flight batch are flushed.
+    run_and_assert_sink_compliance(sink, stream::iter(events), &AWS_SINK_TAGS).await;
```

Comment:
The only assertion that I think _might_ be missing here is one that the sink is not waiting for a batch timeout when shutting down.

Most of the time, batching configs for things like S3 will specify a trigger to flush at a minimum number of events OR a maximum amount of time (e.g. every 1k events or every 5 mins, whichever comes first). This test would probably take a long time if that were happening, but it'd be nice to make it explicit that we are not waiting. A simple way to do that would probably be to explicitly configure that timeout on the sink, and then add a timeout to this statement that is significantly shorter.

Reply:
Great point! Implemented here: https://github.com/vectordotdev/vector/pull/17667/commits/af35d9abcb154098ce198330fa4bf331c3c0ee52